src/main/java/bio/terra/service/snapshot/SnapshotDao.java

```diff
@@ -48,23 +51,86 @@ public SnapshotDao(NamedParameterJdbcTemplate jdbcTemplate,
+     * @param flightId flight id that wants to lock the snapshot
+     * @throws SnapshotLockException if the snapshot is locked by another flight or does not exist
+     */
+    @Transactional(propagation =  Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+    public void lock(String snapshotName, String flightId) {
+        if (flightId == null) {
+            throw new SnapshotLockException("Locking flight id cannot be null");
+        }
+
+        // update the snapshot entry and lock it by setting the flight id
+        String sql = "UPDATE snapshot SET flightid = :flightid " +
+            "WHERE name = :name AND (flightid IS NULL OR flightid = :flightid)";
+        MapSqlParameterSource params = new MapSqlParameterSource()
+            .addValue("name", snapshotName)
+            .addValue("flightid", flightId);
+        int numRowsUpdated = jdbcTemplate.update(sql, params);
+
+        // if no rows were updated, then throw an exception
+        if (numRowsUpdated == 0) {
+            logger.debug("numRowsUpdated=" + numRowsUpdated);
+            throw new SnapshotLockException("Failed to lock the snapshot");
+        }
+    }
+
+    /**
+     * Unlock the snapshot object when finished doing something with it (e.g. delete).
+     * If the snapshot is not locked by this flight, then the method is a no-op. It does not throw an exception in this
+     * case. So, multiple unlocks can succeed with no errors. The method does return a boolean indicating whether any
+     * rows were updated or not. So, callers can decide to throw an error if the unlock was a no-op.
+     * @param snapshotName name of the snapshot to unlock, this is a unique column
+     * @param flightId flight id that wants to unlock the snapshot
+     * @return true if a snapshot was unlocked, false otherwise
+     */
+    @Transactional(propagation =  Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+    public boolean unlock(String snapshotName, String flightId) {
+        // update the snapshot entry to remove the flightid IF it is currently set to this flightid
+        String sql = "UPDATE snapshot SET flightid = NULL " +
+            "WHERE name = :name AND flightid = :flightid";
+        MapSqlParameterSource params = new MapSqlParameterSource()
+            .addValue("name", snapshotName)
+            .addValue("flightid", flightId);
+        int numRowsUpdated = jdbcTemplate.update(sql, params);
+        logger.debug("numRowsUpdated=" + numRowsUpdated);
+        return (numRowsUpdated == 1);
+    }
+
+    /**
+     * Create a new snapshot object and lock it. An exception is thrown if the snapshot already exists.
+     * The correct order to call the SnapshotDao methods when creating a snapshot is: createAndLock, unlock.
+     * @param snapshot the snapshot object to create
+     * @return the id of the new snapshot
+     * @throws DuplicateKeyException if a row already exists with this snapshot name
```

Comment:
I think we should catch that error and throw a DataRepo exception.

Reply:
Moved the conversion of DuplicateKeyException to Invalid[Dataset or Snapshot]Exception from the step class into the daos.